tools/build-toolchain.sh

```diff
@@ -41,32 +78,61 @@ download () {
   if   command_exists wget ; then wget -c  "$1"
   elif command_exists curl ; then curl -LO "$1"
   else
-    echo "Install `wget` or `curl` to download toolchain sources" 1>&2
+    echo "Install 'wget' or 'curl' to download toolchain sources" 1>&2
     return 1
   fi
 }
 
-# Dependency source: Download stage
+echo "Stage: Download and extract dependencies"
 test -f "binutils-$BINUTILS_V.tar.gz" || download "https://ftp.gnu.org/gnu/binutils/binutils-$BINUTILS_V.tar.gz"
+test -d "binutils-$BINUTILS_V"        || tar -xzf "binutils-$BINUTILS_V.tar.gz"
+
 test -f "gcc-$GCC_V.tar.gz"           || download "https://ftp.gnu.org/gnu/gcc/gcc-$GCC_V/gcc-$GCC_V.tar.gz"
+test -d "gcc-$GCC_V"                  || tar -xzf "gcc-$GCC_V.tar.gz" --checkpoint=.100
+
 test -f "newlib-$NEWLIB_V.tar.gz"     || download "https://sourceware.org/pub/newlib/newlib-$NEWLIB_V.tar.gz"
+test -d "newlib-$NEWLIB_V"            || tar -xzf "newlib-$NEWLIB_V.tar.gz"
 
-# Dependency source: Extract stage
-test -d "binutils-$BINUTILS_V" || tar -xzf "binutils-$BINUTILS_V.tar.gz"
-test -d "gcc-$GCC_V"           || tar -xzf "gcc-$GCC_V.tar.gz"
-test -d "newlib-$NEWLIB_V"     || tar -xzf "newlib-$NEWLIB_V.tar.gz"
+# Optional dependency handling
+# Copies the FP libs into GCC sources so they are compiled as part of it
+if [ "$GMP_V" != "" ]; then
+  test -f "gmp-$GMP_V.tar.xz"         || download "https://ftp.gnu.org/gnu/gmp/gmp-$GMP_V.tar.xz"
+  test -d "gmp-$GMP_V"                || tar -xf "gmp-$GMP_V.tar.xz" # note no .gz download file currently available
+  cp -R "gmp-$GMP_V" "gcc-$GCC_V"/gmp #TODO: use symbolic link `ln -s` (cannot get to work)
+fi
+if [ "$MPC_V" != "" ]; then
+  test -f "mpc-$MPC_V.tar.gz"         || download "https://ftp.gnu.org/gnu/mpc/mpc-$MPC_V.tar.gz"
+  test -d "mpc-$MPC_V"                || tar -xzf "mpc-$MPC_V.tar.gz"
+  cp -R "mpc-$MPC_V" "gcc-$GCC_V"/mpc #TODO: use symbolic link `ln -s` (cannot get to work)
+fi
+if [ "$MPFR_V" != "" ]; then
+  test -f "mpfr-$MPFR_V.tar.gz"       || download "https://ftp.gnu.org/gnu/mpfr/mpfr-$MPFR_V.tar.gz"
+  test -d "mpfr-$MPFR_V"              || tar -xzf "mpfr-$MPFR_V.tar.gz"
+  cp -R "mpfr-$MPFR_V" "gcc-$GCC_V"/mpfr #TODO: use symbolic link `ln -s` (cannot get to work)
+fi
+# Certain platforms might require Makefile cross compiling
+if [ "$MAKE_V" != "" ]; then
+  test -f "make-$MAKE_V.tar.gz"       || download "https://ftp.gnu.org/gnu/make/make-$MAKE_V.tar.gz"
+  test -d "make-$MAKE_V"              || tar -xzf "make-$MAKE_V.tar.gz"
+fi
 
-# Compile binutils
+echo "Stage: Compile toolchain"
+
+echo "Compiling binutils-$BINUTILS_V"
 cd "binutils-$BINUTILS_V"
 ./configure \
   --prefix="$INSTALL_PATH" \
   --target=mips64-elf \
   --with-cpu=mips64vr4300 \
   --disable-werror
 make -j "$JOBS"
-make install || sudo make install || su -c "make install"
+make install || sudo make install || su -c "make install" # Perhaps use `checkinstall` instead?!
```

Comment:
One thing to try is to use `make install-strip` instead of `make install`. It should generate much smaller binaries. It should be available for both binutils and GCC.

The dockerfile runs strip manually but I think this is better done here for all kind of usages of the built toolchain.